src/core/src/pattern/matcher.cpp

```diff
@@ -139,13 +140,33 @@ bool Matcher::match_permutation(const OutputVector& pattern_args, const OutputVe
     return true;
 }
 
+// matching arguments in case of `optional`: input values size can be different
+// in this case pattern should cover all possible cases by `optional` pattern
+// against graph with `lost` inputs
+// Operation example: ov::op::v5::NMS
+inline bool are_arguments_misaligned(const std::vector<ov::Output<Node>>& args,
+                                     const std::vector<ov::Output<Node>>& pattern_args) {
+    // 'lost' operation can be defined in the end
+    // try to find them
+    if (pattern_args.size() < args.size()) {
+        return true;
+    }
+    for (size_t i = args.size(); i < pattern_args.size(); ++i) {
+        const auto pattern_in_node_type = pattern_args[i].get_node()->get_type_info();
+        if (!pattern_in_node_type.is_castable(ov::pass::pattern::op::Optional::get_type_info_static())) {
```

Comment:
<img width="415" alt="image" src="https://github.com/openvinotoolkit/openvino/assets/21181928/8fef316e-cea9-4177-8deb-d4f11e34e624">

it looks like this function returns "true" in this case, but expected result is "false".
And I still don't like that we are adding Optional specific to the common place for all patterns.

Reply:
Known limitations were separated to Jira tickets mentioned in the comment